### DISCUSSION TITLE
Cloud Hypervisor being able to boot ISO's

### DISCUSSION THREAD

**Discussion by lonnietc** [cloud-hypervisor/cloud-hypervisor#2864] [Ideas]
Hello All,

Hope that everyone is doing well.

Based upon the things that I have read about Cloud Hypervisor in that it uses the idea of "Crates" to add devices and components to the VMM, I am wondering what it might take to add a crate to allow for it to boot ISO images into the VMM?

Any thoughts on how this might be accomplished?

Also, while still being a bit new to everything, I am curious how these crates are stitched together when adding a new one to the VMM. I guess more reading on Cloud Hypervisor and Rust are needed for me.

Cheers and have a great day,
Lonnie

**Comment by sboeuf** [comment]
Hi! Well if you want to add ISO image format support to Cloud Hypervisor, you would have to implement a new type (something like `IsoDiskFile`) which implements the DiskFile trait. You would also need a secondary type to implement `AsyncIo` trait. Once you have these in place it would be a matter of adding some logic to the VMM code in order to detect if the provided file is an ISO or not.
You can take a look at the different implementations in the `block_util` crate where you'll find `qcow_sync`, `raw_sync`, ...

**Reply by lonnietc** [reply]
This is what I was thinking and if I use the Cloud Hypervisor for the project then I would definitely look into trying to develop some of these crates to contribute back to Cloud Hypervisor as I really like the modular approach that is being used.

Just trying to see how the current code base is laid out and how different components are made to work together now. Mostly in the learning phase at this time, though.

**Comment by rbradford** [comment]
ISO support requires no changes to Cloud Hypervisor. An ISO is so named because it contains an `iso9660` (or `udf`) filesystem (and more commonly these days a hybrid image with a partition table so that you can write it to a USB disk.) From the VMM perspective the image can be handled like any other kind of raw file.

`rust-hypervisor-firmware` does not support mounting `iso9660` file systems however you can probably use the OVMF firmware as described in https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/docs/uefi.md

**Reply by sboeuf** [reply]
Oh that's nice, I never realized ISO was a simple RAW file :)
